File: cpp/src/mc.cpp
#include "localvol/mc.hpp"

#include <cmath>
#include <random>
#include <stdexcept>
#include <vector>

namespace localvol {

namespace {

void validate_mc(const Market& /*market*/, double strike, double expiry, const McSettings& s) {
    if (!std::isfinite(strike) || strike < 0.0) {
        throw std::invalid_argument("mc: strike must be finite and >= 0");
    }
    if (!std::isfinite(expiry) || expiry <= 0.0) {
        throw std::invalid_argument("mc: expiry must be finite and > 0");
    }
    if (s.n_paths < 2) throw std::invalid_argument("mc: n_paths must be >= 2");
    if (s.antithetic && s.n_paths % 2 != 0) {
        throw std::invalid_argument("mc: antithetic sampling requires an even n_paths");
    }
    if (s.n_steps < 1) throw std::invalid_argument("mc: n_steps must be >= 1");
}

VolFn flat_vol_fn(double sigma) {
    if (!std::isfinite(sigma) || sigma <= 0.0) {
        throw std::invalid_argument("mc: flat vol must be finite and > 0");
    }
    return [sigma](double /*k*/, double /*t*/) { return sigma; };
}

/// Mean / standard error over samples; with antithetic on, the samples are
/// pair means (paired payoffs are negatively correlated, so per-path
/// deviations would overstate the error).
McResult estimate(const std::vector<double>& payoff_plus,
                  const std::vector<double>& payoff_minus, bool antithetic, double df) {
    std::vector<double> samples;
    samples.reserve(payoff_plus.size());
    if (antithetic) {
        for (std::size_t i = 0; i < payoff_plus.size(); ++i) {
            samples.push_back(0.5 * df * (payoff_plus[i] + payoff_minus[i]));
        }
    } else {
        for (double p : payoff_plus) samples.push_back(df * p);
    }
    const std::size_t n = samples.size();
    double mean = 0.0;
    for (double v : samples) mean += v;
    mean /= static_cast<double>(n);
    double ss = 0.0;
    for (double v : samples) ss += (v - mean) * (v - mean);
    const double sd = std::sqrt(ss / static_cast<double>(n - 1));  // ddof = 1
    return McResult{mean, sd / std::sqrt(static_cast<double>(n))};
}

McResult european_mc(const Market& market, double strike, double expiry, const VolFn& vol,
                     bool is_call, const McSettings& s) {
    validate_mc(market, strike, expiry, s);
    std::mt19937_64 rng(s.seed);
    std::normal_distribution<double> normal(0.0, 1.0);
    const double dt = expiry / s.n_steps;
    const double sq = std::sqrt(dt);
    const double drift_rq = (market.rate() - market.dividend()) * dt;
    const double x0 = std::log(market.spot());
    const int n_base = s.antithetic ? s.n_paths / 2 : s.n_paths;
    const double phi = is_call ? 1.0 : -1.0;

    std::vector<double> pay_p(static_cast<std::size_t>(n_base));
    std::vector<double> pay_m(s.antithetic ? static_cast<std::size_t>(n_base) : 0);
    for (int p = 0; p < n_base; ++p) {
        double x = x0;
        double xa = x0;
        for (int n = 0; n < s.n_steps; ++n) {
            const double t = n * dt;
            const double lf = market.log_forward(t);
            const double z = normal(rng);
            // Start-of-step lookup at forward log-moneyness X_n - ln F(t_n);
            // the mirrored path recomputes sigma from its own state.
            const double sig = vol(x - lf, t);
            x += drift_rq - 0.5 * sig * sig * dt + sig * sq * z;
            if (s.antithetic) {
                const double siga = vol(xa - lf, t);
                xa += drift_rq - 0.5 * siga * siga * dt - siga * sq * z;
            }
        }
        pay_p[static_cast<std::size_t>(p)] = std::max(phi * (std::exp(x) - strike), 0.0);
        if (s.antithetic) {
            pay_m[static_cast<std::size_t>(p)] = std::max(phi * (std::exp(xa) - strike), 0.0);
        }
    }
    return estimate(pay_p, pay_m, s.antithetic, std::exp(-market.rate() * expiry));
}

McResult barrier_mc(const Market& market, double strike, double barrier, double expiry,
                    const VolFn& vol, const McSettings& s, bool brownian_bridge) {
    validate_mc(market, strike, expiry, s);
    if (!std::isfinite(barrier) || barrier <= 0.0) {
        throw std::invalid_argument("mc: barrier must be finite and > 0");
    }
    if (market.spot() >= barrier) return McResult{0.0, 0.0};  // born knocked out

    std::mt19937_64 rng(s.seed);
    std::normal_distribution<double> normal(0.0, 1.0);
    const double dt = expiry / s.n_steps;
    const double sq = std::sqrt(dt);
    const double drift_rq = (market.rate() - market.dividend()) * dt;
    const double x0 = std::log(market.spot());
    const double b = std::log(barrier);
    const int n_base = s.antithetic ? s.n_paths / 2 : s.n_paths;

    std::vector<double> pay_p(static_cast<std::size_t>(n_base));
    std::vector<double> pay_m(s.antithetic ? static_cast<std::size_t>(n_base) : 0);
    for (int p = 0; p < n_base; ++p) {
        double x = x0, xa = x0;
        double w = 1.0, wa = 1.0;  // survival weights
        for (int n = 0; n < s.n_steps; ++n) {
            const double t = n * dt;
            const double lf = market.log_forward(t);
            const double z = normal(rng);
            {
                const double sig = vol(x - lf, t);
                const double x_new = x + drift_rq - 0.5 * sig * sig * dt + sig * sq * z;
                if (x_new >= b) {
                    w = 0.0;  // discrete knock-out
                } else if (brownian_bridge && w > 0.0) {
                    // Exact bridge crossing probability for x, x_new < b.
                    const double pc = std::exp(-2.0 * (b - x) * (b - x_new) / (sig * sig * dt));
                    w *= 1.0 - pc;
                }
                x = x_new;
            }
            if (s.antithetic) {
                const double sig = vol(xa - lf, t);
                const double x_new = xa + drift_rq - 0.5 * sig * sig * dt - sig * sq * z;
                if (x_new >= b) {
                    wa = 0.0;
                } else if (brownian_bridge && wa > 0.0) {
                    const double pc = std::exp(-2.0 * (b - xa) * (b - x_new) / (sig * sig * dt));
                    wa *= 1.0 - pc;
                }
                xa = x_new;
            }
        }
        pay_p[static_cast<std::size_t>(p)] = w * std::max(std::exp(x) - strike, 0.0);
        if (s.antithetic) {
            pay_m[static_cast<std::size_t>(p)] = wa * std::max(std::exp(xa) - strike, 0.0);
        }
    }
    return estimate(pay_p, pay_m, s.antithetic, std::exp(-market.rate() * expiry));
}

}  // namespace

bool McResult::within(double reference, double n_se) const {
    return std::abs(price - reference) <= n_se * std::max(std_err, 1e-300);
}

McResult price_european_mc(const Market& market, double strike, double expiry, double sigma,
                           bool is_call, const McSettings& settings) {
    return european_mc(market, strike, expiry, flat_vol_fn(sigma), is_call, settings);
}

McResult price_european_mc(const Market& market, double strike, double expiry, const VolFn& vol,
                           bool is_call, const McSettings& settings) {
    if (!vol) throw std::invalid_argument("mc: vol callable must be non-empty");
    return european_mc(market, strike, expiry, vol, is_call, settings);
}

McResult price_up_out_call_mc(const Market& market, double strike, double barrier, double expiry,
                              double sigma, const McSettings& settings, bool brownian_bridge) {
    return barrier_mc(market, strike, barrier, expiry, flat_vol_fn(sigma), settings,
                      brownian_bridge);
}

McResult price_up_out_call_mc(const Market& market, double strike, double barrier, double expiry,
                              const VolFn& vol, const McSettings& settings,
                              bool brownian_bridge) {
    if (!vol) throw std::invalid_argument("mc: vol callable must be non-empty");
    return barrier_mc(market, strike, barrier, expiry, vol, settings, brownian_bridge);
}

}  // namespace localvol

File: rust/src/mc.rs
//! Local-volatility Monte Carlo: log-Euler scheme with antithetic variates.
//!
//! Simulate `X = ln S` on a uniform time grid `t_n = n dt`, `dt = T / N`:
//!
//! ```text
//! X_{n+1} = X_n + (r - q - 1/2 sigma_n^2) dt + sigma_n sqrt(dt) Z_n
//! ```
//!
//! with `sigma_n = sigma(k_n, t_n)` looked up at the *start* of the step,
//! at forward log-moneyness `k_n = X_n - ln F(t_n)` — the same coordinates
//! the Dupire surface uses, so PDE and MC discretise the identical
//! diffusion.  Log-Euler keeps `S > 0` exactly.
//!
//! **Antithetic variates**: paths come in pairs driven by `+Z` and `-Z`
//! (each mirrored path recomputes its own `sigma` from its own state).
//! The estimator averages each pair first, and the standard error is
//! computed over the `n_paths / 2` *pair means* — raw per-path deviations
//! would overstate the error because paired payoffs are negatively
//! correlated.
//!
//! **Up-and-out call**, two monitoring modes:
//!
//! * *Discrete* (`brownian_bridge = false`): a path is dead once a step
//!   ends at or above `ln B`; this overprices the continuous barrier (the
//!   O(sqrt(dt)) discrete-monitoring bias).
//! * *Brownian-bridge correction* (`brownian_bridge = true`, default in the
//!   golden case): for a step from `x0` to `x1`, both below `b = ln B`,
//!   the bridge crossing probability is
//!   `p = exp(-2 (b - x0)(b - x1) / (sigma^2 dt))` and each path carries a
//!   multiplicative survival weight `prod (1 - p)`.  This removes the
//!   leading-order bias with no extra random numbers (the estimator stays
//!   deterministic per seed) and strictly lowers the price versus discrete
//!   monitoring.
//!
//! The RNG is `StdRng` seeded from a fixed `u64`; per the cross-language
//! contract each language uses its own generator, so MC golden comparisons
//! are statistical (tolerances sized at 4 standard errors).

use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use rand_distr::StandardNormal;

use crate::error::{invalid, Result};
use crate::market::Market;
use crate::vol::VolInput;

/// Monte Carlo estimate with its standard error.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct McResult {
    /// Discounted-payoff estimate.
    pub price: f64,
    /// Standard error of the estimate (over pair means when antithetic).
    pub stderr: f64,
}

impl McResult {
    /// True when `reference` lies inside `n_se` standard errors.
    pub fn within(&self, reference: f64, n_se: f64) -> bool {
        (self.price - reference).abs() <= n_se * self.stderr.max(1e-300)
    }
}

/// Simulation settings shared by the MC pricers.
#[derive(Debug, Clone, Copy)]
pub struct McSettings {
    /// Total number of paths (even when `antithetic`).
    pub n_paths: usize,
    /// Number of time steps.
    pub n_steps: usize,
    /// RNG seed (fixed -> deterministic estimate).
    pub seed: u64,
    /// Antithetic variates on/off.
    pub antithetic: bool,
}

impl Default for McSettings {
    fn default() -> Self {
        McSettings {
            n_paths: 20_000,
            n_steps: 100,
            seed: 42,
            antithetic: true,
        }
    }
}

fn validate_mc(strike: f64, expiry: f64, vol: &VolInput, s: &McSettings) -> Result<()> {
    if !strike.is_finite() || strike < 0.0 {
        return Err(invalid(format!("strike must be finite and >= 0, got {strike}")));
    }
    if !expiry.is_finite() || expiry <= 0.0 {
        return Err(invalid(format!("expiry must be finite and > 0, got {expiry}")));
    }
    if s.n_paths < 2 {
        return Err(invalid(format!("n_paths must be >= 2, got {}", s.n_paths)));
    }
    if s.antithetic && s.n_paths % 2 != 0 {
        return Err(invalid("antithetic sampling requires an even n_paths"));
    }
    if s.n_steps < 1 {
        return Err(invalid(format!("n_steps must be >= 1, got {}", s.n_steps)));
    }
    if let Some(sig) = vol.flat() {
        if !sig.is_finite() || sig <= 0.0 {
            return Err(invalid(format!("flat vol must be finite and > 0 for MC, got {sig}")));
        }
    }
    Ok(())
}

/// Pair-mean (antithetic) or plain sample estimator with its standard error.
fn estimate(payoffs: &[f64], antithetic: bool) -> McResult {
    let samples: Vec<f64> = if antithetic {
        let half = payoffs.len() / 2;
        (0..half).map(|i| 0.5 * (payoffs[i] + payoffs[half + i])).collect()
    } else {
        payoffs.to_vec()
    };
    let n = samples.len() as f64;
    let mean = samples.iter().sum::<f64>() / n;
    let var = samples.iter().map(|&p| (p - mean) * (p - mean)).sum::<f64>() / (n - 1.0);
    McResult {
        price: mean,
        stderr: (var / n).sqrt(),
    }
}

/// European vanilla by log-Euler MC (flat vol or Dupire local vol).
pub fn price_european_mc(
    market: &Market,
    strike: f64,
    expiry: f64,
    vol: &VolInput,
    is_call: bool,
    settings: &McSettings,
) -> Result<McResult> {
    validate_mc(strike, expiry, vol, settings)?;
    let mut rng = StdRng::seed_from_u64(settings.seed);
    let dt = expiry / settings.n_steps as f64;
    let sq = dt.sqrt();
    let drift_rq = (market.rate() - market.dividend()) * dt;
    let anti = settings.antithetic;
    let n_base = if anti { settings.n_paths / 2 } else { settings.n_paths };

    let mut x = vec![market.spot().ln(); n_base];
    let mut xa = if anti { x.clone() } else { Vec::new() };
    let mut z = vec![0.0_f64; n_base];
    for n in 0..settings.n_steps {
        let t = n as f64 * dt;
        let lf = market.log_forward(t);
        for zi in z.iter_mut() {
            *zi = rng.sample(StandardNormal);
        }
        for i in 0..n_base {
            let sig = vol.sigma(x[i] - lf, t)?;
            x[i] += drift_rq - 0.5 * sig * sig * dt + sig * sq * z[i];
        }
        if anti {
            for i in 0..n_base {
                let sig = vol.sigma(xa[i] - lf, t)?;
                xa[i] += drift_rq - 0.5 * sig * sig * dt - sig * sq * z[i];
            }
        }
    }

    let phi = if is_call { 1.0 } else { -1.0 };
    let df = (-market.rate() * expiry).exp();
    let payoff = |xt: f64| df * (phi * (xt.exp() - strike)).max(0.0);
    let payoffs: Vec<f64> = x.iter().chain(xa.iter()).map(|&xt| payoff(xt)).collect();
    Ok(estimate(&payoffs, anti))
}

/// Up-and-out call under local (or flat) vol, optional Brownian-bridge
/// correction.  Requires `S0 < barrier` (otherwise the option is born
/// knocked out and worth exactly 0, returned with zero standard error).
pub fn price_up_out_call_mc(
    market: &Market,
    strike: f64,
    barrier: f64,
    expiry: f64,
    vol: &VolInput,
    settings: &McSettings,
    brownian_bridge: bool,
) -> Result<McResult> {
    validate_mc(strike, expiry, vol, settings)?;
    if !barrier.is_finite() || barrier <= 0.0 {
        return Err(invalid(format!("barrier must be finite and > 0, got {barrier}")));
    }
    if market.spot() >= barrier {
        return Ok(McResult { price: 0.0, stderr: 0.0 });
    }

    let mut rng = StdRng::seed_from_u64(settings.seed);
    let dt = expiry / settings.n_steps as f64;
    let sq = dt.sqrt();
    let drift_rq = (market.rate() - market.dividend()) * dt;
    let b = barrier.ln();
    let anti = settings.antithetic;
    let n_base = if anti { settings.n_paths / 2 } else { settings.n_paths };

    // One shared normal block so the antithetic pass reuses the same draws.
    let z_all: Vec<Vec<f64>> = (0..settings.n_steps)
        .map(|_| (0..n_base).map(|_| rng.sample(StandardNormal)).collect())
        .collect();

    let run_block = |sign: f64| -> Result<Vec<f64>> {
        let mut x = vec![market.spot().ln(); n_base];
        let mut weight = vec![1.0_f64; n_base];
        for (n, z) in z_all.iter().enumerate() {
            let t = n as f64 * dt;
            let lf = market.log_forward(t);
            for i in 0..n_base {
                let sig = vol.sigma(x[i] - lf, t)?;
                let x_new = x[i] + drift_rq - 0.5 * sig * sig * dt + sign * sig * sq * z[i];
                if x_new >= b {
                    weight[i] = 0.0;
                } else if brownian_bridge && weight[i] > 0.0 {
                    // P[bridge from x to x_new crosses b], both below b.
                    let p = (-2.0 * (b - x[i]) * (b - x_new) / (sig * sig * dt)).exp();
                    weight[i] *= 1.0 - p;
                }
                x[i] = x_new;
            }
        }
        Ok((0..n_base)
            .map(|i| weight[i] * (x[i].exp() - strike).max(0.0))
            .collect())
    };

    let mut payoffs = run_block(1.0)?;
    if anti {
        payoffs.extend(run_block(-1.0)?);
    }
    let df = (-market.rate() * expiry).exp();
    for p in payoffs.iter_mut() {
        *p *= df;
    }
    Ok(estimate(&payoffs, anti))
}
